presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewRewriteQueryShapeValidator.java

```diff
@@ -0,0 +1,89 @@
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.analyzer;
+
+import com.facebook.presto.sql.tree.DefaultTraversalVisitor;
+import com.facebook.presto.sql.tree.FunctionCall;
+import com.facebook.presto.sql.tree.GroupBy;
+import com.facebook.presto.sql.tree.QualifiedName;
+import com.facebook.presto.sql.tree.QuerySpecification;
+import com.google.common.collect.ImmutableSet;
+
+import java.util.Optional;
+import java.util.Set;
+
+import static java.lang.String.format;
+
+/**
+ * Validates that the {@link QuerySpecification} provided to
+ * {@link MaterializedViewRewriteQueryShapeValidator#validate}
+ * is able to be rewritten from a backing materialized view. This does not guarantee the
+ * QuerySpecification has a compatible materialized view to be rewritten with, only that
+ * its shape is not prohibitive to the optimization.
+ */
+public final class MaterializedViewRewriteQueryShapeValidator
+{
+    private MaterializedViewRewriteQueryShapeValidator() {};
+
+    static Optional<String> validate(QuerySpecification querySpecification)
+    {
+        return new Visitor().validateMaterializedViewOptimizationQueryShape(querySpecification);
+    }
+
+    static class Visitor
+            extends DefaultTraversalVisitor<Void, Void>
+    {
+        private static final QualifiedName MIN = QualifiedName.of("MIN");
+        private static final QualifiedName MAX = QualifiedName.of("MAX");
+        private static final QualifiedName SUM = QualifiedName.of("SUM");
+        private static final QualifiedName COUNT = QualifiedName.of("COUNT");
+        private static final Set<QualifiedName> SUPPORTED_FUNCTION_CALLS = ImmutableSet.of(MIN, MAX, SUM, COUNT);
+
+        private Optional<String> errorMessage = Optional.empty();
+        boolean hasGroupBy;
+
+        public Optional<String> validateMaterializedViewOptimizationQueryShape(QuerySpecification querySpecification)
+        {
+            if (querySpecification.getHaving().isPresent()) {
+                errorMessage = Optional.of("Query shape invalid: HAVING is not supported for materialized view optimizations");
+            }
+            if (querySpecification.getLimit().isPresent()) {
+                errorMessage = Optional.of("Query shape invalid: LIMIT is not supported for materialized view optimizations");
+            }
+            else {
+                process(querySpecification);
+            }
+
+            // TODO: add a check requiring GROUP BY or WHERE to be present
+            return errorMessage;
+        }
+
+        @Override
+        protected Void visitGroupBy(GroupBy node, Void context)
+        {
+            hasGroupBy = true;
+            return super.visitGroupBy(node, context);
+        }
+
+        @Override
+        protected Void visitFunctionCall(FunctionCall node, Void context)
+        {
+            QualifiedName name = node.getName();
+            if (!SUPPORTED_FUNCTION_CALLS.contains(name)) {
```

Comment:
Inline `node.getName()`

Reply:
Using name as intended in the error message instead. WDYT?